### PR TITLE
Add keepEndOfOptions flag to std.getopt

### PR DESCRIPTION
**Pull request by mrkline**
Currently, the endOfOptions separator ("--" by default) is stripped from args by getopt.  At times it may be advantageous to keep this separator, such as when your program is forwarding command line arguments to some other program.

Without this, hacks such as

```D
auto remaining = args.find(endOfOptions);
args = args[0 .. $ - remaining.length];
getopt(args, opts);
args = args ~ remaining;
```

are needed to achieve the desired behavior.

### CODE REVIEW

File: std/getopt.d [hunk 1]
```diff
@@ -1102,6 +1108,23 @@ unittest
     getopt(args, "fb1", &fb1, "fb2", &fb2, "tb1", &tb1);
     assert(fb1 && fb2 && !tb1);
 
+    // test keepEndOfOptions
+
+    args = (["program.name", "--foo", "nonoption", "--bar", "--", "--baz"]).dup;
```
Comment (mrkline) [inline std/getopt.d]: As an aside, I'm fairly new to Phobos development and must ask - what is the rationale behind the `(<array>).dup` pattern in these unit tests?

Comment (JakobOvrum) [inline std/getopt.d]: Looks pretty bonkers to me. Why allocate twice?

Comment (mrkline) [inline std/getopt.d]: Agreed, but all the existing tests in that `unittest {}` block do the same.

Comment (JakobOvrum) [inline std/getopt.d]: I'm guessing the code is really old and there were circumstances calling for this awkwardness. In this PR it's probably nice to keep it consistent with what is there; it should rather be fixed wholesale in a different PR.

Comment (schveiguy) [inline std/getopt.d]: Yeah, it's pretty old. But I think it's far from consistent. You can easily find cases where the `dup` was forgotten. I don't think there was ever a reason to have the `dup`.

In any case, you can remove the `dup` or not, it won't really change anything.

Comment (mrkline) [inline std/getopt.d]: Alright. Later I'll probably open a PR that gets rid of them elsewhere.

**Comment by yebblies** [comment]
I've approved you on the autotester. https://auto-tester.puremagic.com/pulls.ghtml?projectid=1

**Comment by JakobOvrum** [comment]
Implementation LGTM, but I've hardly used std.getopt and I'm in no position to evaluate the usefulness of this feature. The argument presented in the OP looks good, but hopefully someone more familiar with the subject can evaluate it.

**Comment by burner** [comment]
can you explain what expect to be in args after you passed it into getopt. 
Is "--" an cmd option for the other program you where talking about.

**Comment by quickfur** [comment]
LGTM

**Comment by quickfur** [comment]
@burner The unittest answers your question re: what is expected to be in args after `getopt` returns.

**Comment by quickfur** [comment]
Auto-merge toggled on

**Comment by quickfur** [comment]
Auto-merge toggled off

**Comment by quickfur** [comment]
Auto-merge toggled on

**Comment by quickfur** [comment]
Gah, pushed the button once too many times.

**Comment by burner** [comment]
yeah, I got will going home
